_posts/2025-08-05-fcmp++-contest-final.md

```diff
@@ -0,0 +1,74 @@
+
+`helioselene` is a library implementing two curves, "Helios" and "Selene," which form a curve cycle towering Ed25519. Using a "tower cycle" over Ed25519 allows Monero to re-use the existing anonymity set while upgrading to FCMP++. This library is a critical component used in all aspects of FCMP++. Optimizing this library will therefore benefit all user-facing components of FCMP++ (daemon sync, wallet sync, and transaction construction).
+
+Although not part of the official contest rules, we decided to award a 2nd place prize to Rafael, since we unanimously agreed Rafael's submission was 2nd best. Had we not received a submission from Lederstrumpf, we would have had an excellent submission to fall back on [here](https://github.com/rafael-xmr/fcmp-plus-plus-optimization-competition-private/pull/5). As such, we sought to reward Rafael's strong effort.
+
+We received a number of quality submissions to this contest, and want to thank all contestants for participating! You can read more on our rationale for selecting the winner [here](https://github.com/j-berman/fcmp-plus-plus-optimization-competition/blob/main/docs/helioselene-decision.pdf).
+
+## What we learned
+
+### Form stricter bounds on exactly what code can be modified
+
+Some code in a repo required for tests / benchmarks to run might not be in scope of the competition. Be as clear about which code is out of scope as possible. Explicitly tell contestants that modifiying certain code will be disqualifying.
+
+### Form stricter API requirements
+
+Similar to the previous point, be crystal clear about the requirements the API of the code must take on and if it should even be allowed to be modified. Preferably provide downstream code that the API must compile and function correctly with.
+
+### Present concrete architectures and toolchains for which the code must run in constant-time
+
+Code runs in constant-time or doesn't on specific architectures and toolchains. For example, LLVM to x86 generates vatiable-time code for u128 arithmetic in Rust, but constant-time code on x86_64. Give contestants examples of specific toolchains and architectures for which the code must compile to constant-time.
+
+### Allocate time for review and 2nd phase
+
+No matter how well you think you define the competition rules, there may be scenarios which crop up where there was widespread confusion about the rules. There may be multiple submisisons which have a lot of good work in them, but are in the gray area of validity. A review phase allows judges to clarify rules with contestants and refine submissions, and planning for this ahead of time avoids misunderstandings and hurt feelings.
```

Comment:
```suggestion
No matter how well you think you define the competition rules, there may be scenarios which crop up where there was widespread confusion about the rules. There may be multiple submissions which have a lot of good work in them, but are in the gray area of validity. A review phase allows judges to clarify rules with contestants and refine submissions, and planning for this ahead of time helps to avoid misunderstandings.
```

"Hurt feelings" might hurt feelings :P